plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java

```diff
@@ -295,21 +313,32 @@ private void doVacuum(
                     retainedUnknownFiles++;
                     continue;
                 }
-
                 log.debug("[%s] deleting file [%s] with modification time %s", queryId, location, modificationTime);
-                filesToDelete.add(entry.location());
-                if (filesToDelete.size() == DELETE_BATCH_SIZE) {
-                    fileSystem.deleteFiles(filesToDelete);
-                    removedFiles += filesToDelete.size();
-                    filesToDelete.clear();
-                }
-            }
 
-            if (!filesToDelete.isEmpty()) {
-                fileSystem.deleteFiles(filesToDelete);
-                removedFiles += filesToDelete.size();
+                Location fileLocation = Location.of(location);
+                TrinoInputFile inputFile = fileSystem.newInputFile(fileLocation);
+                filesToDelete.add(inputFile);
+                filesToDeleteSize += inputFile.length();
             }
+            long readVersion = handle.getReadVersion();
+            logVacuumStart(handle.location(), session, readVersion, filesToDelete.size(), filesToDeleteSize);
+            int totalFilesToDelete = filesToDelete.size();
+            int batchCount = (int) Math.ceil((double) totalFilesToDelete / DELETE_BATCH_SIZE);
+            String status = "FAILED";
+            try {
+                for (int batchNumber = 0; batchNumber < batchCount; batchNumber++) {
+                    int start = batchNumber * DELETE_BATCH_SIZE;
+                    int end = Math.min(start + DELETE_BATCH_SIZE, totalFilesToDelete);
 
+                    List<TrinoInputFile> batch = filesToDelete.subList(start, end);
+                    fileSystem.deleteFiles(batch.stream().map(TrinoInputFile::location).collect(toImmutableList()));
+                }
+                status = "COMPLETED";
+            }
+            finally {
+                int numVacuumedDirectories = Set.of(filesToDelete.stream().map(f -> f.location().parentDirectory())).size();
+                logVacuumEnd(handle.location(), session, readVersion, filesToDelete.size(), numVacuumedDirectories, status);
```

Comment:
Instead of `filesToDelete.size()` we should count actually deleted files in case of exception in the middle of the process.